stdlib/LinearAlgebra/src/qr.jl

```diff
@@ -737,6 +737,42 @@ function *(adjA::Adjoint{<:Any,<:StridedVecOrMat}, adjQ::Adjoint{<:Any,<:Abstrac
     return rmul!(Ac, adjoint(convert(AbstractMatrix{TAQ}, Q)))
 end
 
+### mul!
+function check_dimensions(C::StridedVecOrMat{T}, tA::AbstractChar, tB::AbstractChar,
+    A::Union{StridedVecOrMat{T}, AbstractQ{T}},
+    B::Union{AbstractQ{T}, StridedVecOrMat{T}}) where T<:BlasFloat
+    # Auxiliary function used in mul! with StridedVecOrMat and Q matrices from QR decompositions
+    mA, nA = lapack_size(tA, A)
+    mB, nB = lapack_size(tB, B)
+
+    if nA != mB
+        throw(DimensionMismatch("A has dimensions ($mA,$nA) but B has dimensions ($mB,$nB)"))
+    end
+    if lapack_size('N', C) != (mA, nB)
+        throw(DimensionMismatch("C has dimensions $(size(C)), should have ($mA,$nB)"))
+    end
+end
+
+function mul!(C::StridedVecOrMat{T}, Q::AbstractQ{T}, B::StridedVecOrMat{T}) where T<:BlasFloat
+    check_dimensions(C, 'N', 'N', Q, B)
```

Comment:
Is it necessary to check the dimensions here? Both `copyto!` and two-argument `lmul!` have checks so I wouldn't think it would be needed here.

Reply:
With `StridedVecOrMat`s of `BlasFloat`s, `lmul!` calls `LAPACK.gemqrt!` or `LAPACK.ormqr!` directly. So it depends on whether these assume correct sizes already.

On a different note, there is also `function lmul!(A::QRPackedQ, B::AbstractVecOrMat)`. Should we then also have the proposed `mul!` methods here for abstract vectors or matrices?

Reply:
`LAPACK.gemqrt!` and `LAPACK.ormqr!` seem to check dimensions, but `copyto!(A, B)` only checks that `length(A) >= length(B)' which is less strict that the checks I have.

Reply:
Yes, thanks, I think that we should also support `AbstractVecOrMat`. On a related note, I cannot understand thought why `rmul!` requires `StridedMatrix` while `lmul!` does not, i.e.
```julia
function lmul!(A::QRPackedQ, B::AbstractVecOrMat)
function rmul!(A::StridedMatrix, adjQ::Adjoint{<:Any,<:QRPackedQ})
```

Reply:
If one wanted to have stricter checks, one could use `copy!` instead of `copyto!`, which requires equal axes. I couldn't find any use of `copy!` in `qr.jl` though. That sounds like there's no need for additional size checks then?

Reply:
Hm, it occurred to me that for mutating `*mul!` functions you need memory to write into. So it could be that the destination arrays need to be strided, but for 3-arg `mul!`, the source array could be abstract. Does that make sense for a general rule?

Reply:
@dkarrasch thanks I wasn't aware of `copy!`. I will use that and remove the additional size checks.

Reply:
Well, I thought that since both the LAPACK functions and the generic Julian `*mul!` functions do size checks, it would suffice to leave it to them and to use the more "generous" `copyto!` (as suggested by @andreasnoack), independently of whether you go the LAPACK or the generic Julia route. Basically, leave everything as you have, but kick out size checks.